android/java/org/chromium/chrome/browser/app/domain/KeyringModel.java

```diff
@@ -301,73 +380,92 @@ public void onChanged(Boolean success) {
         }
     }
 
+    private void finalizeWalletRestoration(
+            @NonNull final String password,
+            @NonNull final String recoveryPhrase,
+            final boolean legacyRestoreEnabled,
+            @NonNull final Set<NetworkInfo> selectedNetworks,
+            @NonNull final Callbacks.Callback1<Boolean> callback) {
+        assertOnUiThread();
+        mKeyringService.restoreWallet(
+                recoveryPhrase,
+                password,
+                legacyRestoreEnabled,
+                result -> createAccounts(result, selectedNetworks, callback));
+    }
+
     private void finalizeWalletCreation(
             @NonNull final String password,
             @NonNull final Set<NetworkInfo> selectedNetworks,
             @NonNull final Callbacks.Callback1<String> callback) {
         assertOnUiThread();
         mKeyringService.createWallet(
                 password,
-                recoveryPhrases -> {
-                    final Set<NetworkInfo> createAccounts = new HashSet<>();
+                recoveryPhrase -> createAccounts(recoveryPhrase, selectedNetworks, callback));
+    }
+
+    private <T> void createAccounts(
+            final T result,
+            @NonNull final Set<NetworkInfo> selectedNetworks,
+            @NonNull final Callbacks.Callback1<T> callback) {
+        final Set<NetworkInfo> createAccounts = new HashSet<>();
+
+        for (NetworkInfo networkInfo : selectedNetworks) {
+            // Create Bitcoin and Filecoin accounts if they have been selected.
+            if (networkInfo.coin == CoinType.BTC || networkInfo.coin == CoinType.FIL) {
+                createAccounts.add(networkInfo);
+            }
+        }
+
+        if (createAccounts.isEmpty()) {
+            callback.call(result);
+        } else {
+            final MutableLiveData<Boolean> createAccountsLiveData = new MutableLiveData<>();
+
+            createAccountsLiveData.observeForever(
```

Comment:
should we use LiveDataUtil.observeOnce here ?

Reply:
Unfortunately it didn't work here as I needed to observe the `createAccountsLiveData` not only once but for the entire time the accounts are emitted by `mAccountInfos` (a few lines below). When the number of emitted accounts matches the size of `createAccounts` then and only then we can suspend the observer.